Add @attr to cache function return values

fingolfin opened the pull request:
Using our attributes system, it is easy to cache the return value
of expensive functions: Just turn

    function myattr(obj::Foo)
        # ... expensive computation
        return result::fmpz
    end

into something like this (provided that `Foo` supports attribute storing)

    function myattr(obj::Foo)
      return get_attribute!(obj, :myattr) do
        # ... expensive computation
        return result::fmpz
      end
    end

However, doing this repeatedly quickly gets tedious. It also comes with a
catch: it prevents type inference on the return value of `myattr`: while in
the first example `Base.return_types(myattr)` produces `[fmpz]`, in the second
example it yields `[Any]`. This can be resolved by manually adding a suitable type
assertion on the return value of `myattr`, but that's easy to forget.

The `@attr` macro resolves this and more. With it, you can write the following
code to get the same caching as in the second example, but with a correctly
inferred return type:

    @attr function myattr(obj::Foo)
        # ... expensive computation
        return result::fmpz
    end

Under the hood, this works by renaming the function doing the actual
computation, then calling `Base.return_types` to determine its return type(s).
Then a wrapper function under the original name is created which invokes
`get_attribute!` suitably and applies a type assertion based on the return
type(s) that were just computed, i.e., basically what was sketched above, but
all done automatically.


---

This is not quite done, there are a few TODOs and I need to test this more, and perhaps the macro should have a better name, and perhaps I should generate `has_attribute` helpers, etc. But it's already good for CI testing and getting feedback.

CC @thofma @HereAround

thofma commented:
1. I am not sure I understand the "catch". The `::fmpz` just needs to move a bit and everything is fine?
2. MacroTools.jl is famous for being very slow. How are load times effected? I mean, once `@attr` is springled at hundreds of functions.
3. I am not a big fan of using `Base.return_types`, as it might go away at any time.

My biggest concern is 2.

P.S.: Since in the example there is always an explicit return type `fmpz` flying around, couldn't this just be another argument to the `get_attribute!` function?

HereAround commented:
@fingolfin Thank you for pushing this. I think the syntax would indeed greatly simply the caching.

I just came to wonder about logic deductions. When we compute an attribute or property, then we can use the result of this to infer - at least sometimes - other attributes and properties. For example:
1. Compute `isaffine(toric_variety)`.
2. Find that this is `true`.
3. Consequence: The variety is not complete and not projective.

For ease of maintenance, one could write a central function (say `logic(v::AbstractToricVariety)`) which conducts these deductions. The key question is then likely when/how to call this function. I am wondering if it makes sense to integrate such a call into the @attr wrapper.

I imagine something like this:
> 
> ```
> @logicAndattr function myattr(obj::Foo)
>     # ... expensive computation
>     return result::fmpz
> end
> ```
> 
which acts as follows:
1. Set the attribute `myattr` for the object `obj` of type `Foo`, i.e. just as @attr is intended.
2. Once complete, issues `logic(obj)`, which then performs the logical deductions.

If the 2nd step is not desired/beneficial, one could fall back to use @attr instead.

What do you think? Is this reasonable? Technically possible?

fieker commented:
On Mon, Jan 03, 2022 at 01:19:39PM -0800, Martin Bies wrote:
> @fingolfin Thank you for pushing this. I think the syntax would indeed greatly simply the caching.
> 
> I just came to wonder about logic deductions. When we compute an attribute or property, then we can use the result of this to infer - at least sometimes - other attributes and properties. For example:
> 1. Compute `isaffine(toric_variety)`.
> 2. Find that this is `true`.
> 3. Consequence: The variety is not complete and not projective.
> 
> For ease of maintenance, one could write a central function (say `logic(v::AbstractToricVariety)`) which conducts these deductions. The key question is then likely when/how to call this function. I am wondering if it makes sense to integrate such a call into the @attr wrapper.
> 
> I imagine something like this:
> > 
> > ```
> > @logicAndattr function myattr(obj::Foo)
> >     # ... expensive computation
> >     return result::fmpz
> > end
> > ```
> > 
> which acts as follows:
> 1. Set the attribute `myattr` for the object `obj` of type `Foo`, i.e. just as @attr is intended.
> 2. Once complete, issues `logic(obj)`, which then performs the logical deductions.
> 
> If the 2nd step is not desired/beneficial, one could fall back to use @attr instead.
> 
> What do you think? Is this reasonable? Technically possible?
> 
This all depends: 
 how many logicals do you want to test/set?
the current proposed interface would require you to scan all possible
attributes to decide if any additional ones can be derived. This can be
tedious.
Of course, logics could also get the attribute as an argument, even as a
val-type so dispatch could settle this.
However, what is the difference to, in your example, to implement
 iscomplete
   if isaffine is known and true
     return false
   else
     do s.th.
 isprojective
   if isaffine is known and true
     return false
   else
     do s.th.
What would be the advantage of the logic function over this?     

> -- 
> Reply to this email directly or view it on GitHub:
> https://github.com/Nemocas/AbstractAlgebra.jl/pull/1092#issuecomment-1004366268
> You are receiving this because you are subscribed to this thread.
> 
> Message ID: ***@***.***>

HereAround commented:
> This all depends: how many logicals do you want to test/set? the current proposed interface would require you to scan all possible attributes to decide if any additional ones can be derived. This can be tedious. Of course, logics could also get the attribute as an argument, even as a val-type so dispatch could settle this. However, what is the difference to, in your example, to implement iscomplete if isaffine is known and true return false else do s.th. isprojective if isaffine is known and true return false else do s.th. What would be the advantage of the logic function over this?
> […](#)
> -- Reply to this email directly or view it on GitHub: [#1092 (comment)](https://github.com/Nemocas/AbstractAlgebra.jl/pull/1092#issuecomment-1004366268) You are receiving this because you are subscribed to this thread. Message ID: ***@***.***>

Agreed.

I am not yet sure how many quantities I would like to test and how frequently this should be done. But the hope would be to use logic for all quantities in the toric varieties. Currently, we have about 10 properties for toric varieties and order 30 attributes. More properties/attributes for toric divisor and line bundles.

(And, at least in principle, one could envision to extend the caching and logic deduction to all of Oscar, one the architecture is agreed upon.)

Suppose that I follow your proposal and check the order 40 quantities for toric varieties for logic. Arguably, this then becomes messy pretty quickly as I have to modify many functions. While certainly possible, I came to think that a central "logic" function would be easier to maintain/keep the code easy to read.

It would make sense (to me) to have one logic function for toric varieties, one for toric divisors and yet another one for toric line bundles. Also, as you suggested, one could give a string as additional argument (e.g. "affine" in the above example), so that only deduction are triggered based on the knowledge of that very property/attribute.

I was uncertain when best to trigger this logic function. Too often and we lose performance and seldom do we see benefits if we almost never trigger it. By checking immediately after obtaining a new cached property seems (at least on first sight) not unreasonable. Hence, my question whether or not it makes sense to integrate such an approach into the @attr wrapper.

fingolfin commented:
Regarding "boolean attributes" (which in GAP called "properties") and logical implications between them (for which GAP has some machinery of tracking them): I do have some ideas on that, and I think we should talk (more) about it (I did talk a bit about it with @fieker today); but I'd like to do this separately from this PR, one step at a time, esp. since such a machinery can also easily introduce severe costs (as we learned the hard way in GAP) if not designed very carefully. It can be overcome, but it'll always be a tradeoff. Anyway, enough on that for now and here...

@thofma wrote:

> 1. I am not sure I understand the "catch". The `::fmpz` just needs to move a bit and everything is fine?

Yes it does (actually, I am surprised, because when I made experiments with this earlier, it didn't quite do the right thing -- I must have screwed up my earlier tests, ooops!)

That said, not that existing code usually has no such type assertions (and often doesn't need it, thanks to type inference), I put it there for illustrative purposes. I just thought it would be nice to have this "for free" just like it is (often) "for free" in regular methods.


> 2. MacroTools.jl is famous for being very slow. How are load times effected? I mean, once `@attr` is springled at hundreds of functions.

I wasn't aware of this, and could conduct some experiments. But note that we only use it in a macro, so it should only affect precompilation time, not load time (at least not more than inserting the corresponding code manually would)

> 3. I am not a big fan of using `Base.return_types`, as it might go away at any time.

Is that really like? If that were to happen, we'd have to ditch the nice automatic determination of return types, but that would be an easy change to the macro. Of course we'd then have to add lots of manual, explicit type assertions... 

But, fair enough.

> 
> My biggest concern is 2.
> 
> P.S.: Since in the example there is always an explicit return type `fmpz` flying around, couldn't this just be another argument to the `get_attribute!` function?

It could be, though I don't think there's a win doing that over just adding the type assertions; as you pointed out.

A variant of this would be to have this macro but allow passing the type to it as an (optional?) argument. So you might write:
```julia
@attr fmpz function myattr(obj::Foo)
    # ... expensive computation
    return result
end
```

thofma commented:
> I wasn't aware of this, and could conduct some experiments. But note that we only use it in a macro, so it should only affect precompilation time, not load time (at least not more than inserting the corresponding code manually would)

I realized that I mixed up MacroTools with IRTools, my bad. Sorry MacroTools. (It will still mean inferring large parts of the Oscar code base during recompilation time, but this might be fine?).

So no objections from my part to adding this.

thofma commented on src/Attributes.jl:
```diff
@@ -219,3 +222,123 @@ function set_attribute!(G::Any, attr::Symbol, value::Any)
+julia> @attributes mutable struct Foo
+           x::Int
+           Foo(x::Int) = new(x)
+       end
```
```suggestion
       end;
```

thofma commented on src/Attributes.jl:
```diff
@@ -219,3 +222,123 @@ function set_attribute!(G::Any, attr::Symbol, value::Any)
+julia> @attr function myattr(obj::Foo)
+                println("Performing expensive computation")
+                return factorial(obj.x)
+             end
```
```suggestion
             end;
```

fingolfin commented:
@thofma pointed out that adding this could cause a precompilation slowdown because it'd force Julia to run type inference on more code (which could also be a boon during runtime, hard to say). I'll run benchmarks on this and report.

fingolfin commented:
I wrote a PR to change Oscar.jl to use `@attr` in most places that use `get_attribute!` right now, and... it wasn't encouraging :-(
```
signal (11): Segmentation fault: 11
in expression starting at REPL[1]:1
jl_deserialize_value at /Users/mhorn/Applications/Julia-1.6.5.app/Contents/Resources/julia/lib/julia/libjulia-internal.1.dylib (unknown line)
jl_deserialize_value_array at /Users/mhorn/Applications/Julia-1.6.5.app/Contents/Resources/julia/lib/julia/libjulia-internal.1.dylib (unknown line)
jl_deserialize_value at /Users/mhorn/Applications/Julia-1.6.5.app/Contents/Resources/julia/lib/julia/libjulia-internal.1.dylib (unknown line)
jl_deserialize_value at /Users/mhorn/Applications/Julia-1.6.5.app/Contents/Resources/julia/lib/julia/libjulia-internal.1.dylib (unknown line)
jl_deserialize_value_array at /Users/mhorn/Applications/Julia-1.6.5.app/Contents/Resources/julia/lib/julia/libjulia-internal.1.dylib (unknown line)
jl_deserialize_value at /Users/mhorn/Applications/Julia-1.6.5.app/Contents/Resources/julia/lib/julia/libjulia-internal.1.dylib (unknown line)
jl_deserialize_value at /Users/mhorn/Applications/Julia-1.6.5.app/Contents/Resources/julia/lib/julia/libjulia-internal.1.dylib (unknown line)
jl_deserialize_value at /Users/mhorn/Applications/Julia-1.6.5.app/Contents/Resources/julia/lib/julia/libjulia-internal.1.dylib (unknown line)
jl_deserialize_value at /Users/mhorn/Applications/Julia-1.6.5.app/Contents/Resources/julia/lib/julia/libjulia-internal.1.dylib (unknown line)
jl_deserialize_value at /Users/mhorn/Applications/Julia-1.6.5.app/Contents/Resources/julia/lib/julia/libjulia-internal.1.dylib (unknown line)
jl_deserialize_value at /Users/mhorn/Applications/Julia-1.6.5.app/Contents/Resources/julia/lib/julia/libjulia-internal.1.dylib (unknown line)
jl_deserialize_value at /Users/mhorn/Applications/Julia-1.6.5.app/Contents/Resources/julia/lib/julia/libjulia-internal.1.dylib (unknown line)
jl_deserialize_value at /Users/mhorn/Applications/Julia-1.6.5.app/Contents/Resources/julia/lib/julia/libjulia-internal.1.dylib (unknown line)
jl_deserialize_value_array at /Users/mhorn/Applications/Julia-1.6.5.app/Contents/Resources/julia/lib/julia/libjulia-internal.1.dylib (unknown line)
jl_deserialize_value at /Users/mhorn/Applications/Julia-1.6.5.app/Contents/Resources/julia/lib/julia/libjulia-internal.1.dylib (unknown line)
jl_deserialize_value at /Users/mhorn/Applications/Julia-1.6.5.app/Contents/Resources/julia/lib/julia/libjulia-internal.1.dylib (unknown line)
jl_deserialize_value at /Users/mhorn/Applications/Julia-1.6.5.app/Contents/Resources/julia/lib/julia/libjulia-internal.1.dylib (unknown line)
jl_deserialize_value at /Users/mhorn/Applications/Julia-1.6.5.app/Contents/Resources/julia/lib/julia/libjulia-internal.1.dylib (unknown line)
jl_deserialize_value at /Users/mhorn/Applications/Julia-1.6.5.app/Contents/Resources/julia/lib/julia/libjulia-internal.1.dylib (unknown line)
jl_deserialize_value at /Users/mhorn/Applications/Julia-1.6.5.app/Contents/Resources/julia/lib/julia/libjulia-internal.1.dylib (unknown line)
jl_deserialize_value at /Users/mhorn/Applications/Julia-1.6.5.app/Contents/Resources/julia/lib/julia/libjulia-internal.1.dylib (unknown line)
jl_deserialize_value at /Users/mhorn/Applications/Julia-1.6.5.app/Contents/Resources/julia/lib/julia/libjulia-internal.1.dylib (unknown line)
jl_deserialize_value at /Users/mhorn/Applications/Julia-1.6.5.app/Contents/Resources/julia/lib/julia/libjulia-internal.1.dylib (unknown line)
jl_deserialize_value at /Users/mhorn/Applications/Julia-1.6.5.app/Contents/Resources/julia/lib/julia/libjulia-internal.1.dylib (unknown line)
jl_deserialize_value at /Users/mhorn/Applications/Julia-1.6.5.app/Contents/Resources/julia/lib/julia/libjulia-internal.1.dylib (unknown line)
jl_deserialize_value at /Users/mhorn/Applications/Julia-1.6.5.app/Contents/Resources/julia/lib/julia/libjulia-internal.1.dylib (unknown line)
jl_deserialize_value at /Users/mhorn/Applications/Julia-1.6.5.app/Contents/Resources/julia/lib/julia/libjulia-internal.1.dylib (unknown line)
jl_deserialize_value at /Users/mhorn/Applications/Julia-1.6.5.app/Contents/Resources/julia/lib/julia/libjulia-internal.1.dylib (unknown line)
jl_deserialize_value at /Users/mhorn/Applications/Julia-1.6.5.app/Contents/Resources/julia/lib/julia/libjulia-internal.1.dylib (unknown line)
jl_deserialize_value at /Users/mhorn/Applications/Julia-1.6.5.app/Contents/Resources/julia/lib/julia/libjulia-internal.1.dylib (unknown line)
jl_deserialize_value at /Users/mhorn/Applications/Julia-1.6.5.app/Contents/Resources/julia/lib/julia/libjulia-internal.1.dylib (unknown line)
jl_deserialize_value at /Users/mhorn/Applications/Julia-1.6.5.app/Contents/Resources/julia/lib/julia/libjulia-internal.1.dylib (unknown line)
jl_deserialize_value at /Users/mhorn/Applications/Julia-1.6.5.app/Contents/Resources/julia/lib/julia/libjulia-internal.1.dylib (unknown line)
jl_deserialize_value_array at /Users/mhorn/Applications/Julia-1.6.5.app/Contents/Resources/julia/lib/julia/libjulia-internal.1.dylib (unknown line)
_jl_restore_incremental at /Users/mhorn/Applications/Julia-1.6.5.app/Contents/Resources/julia/lib/julia/libjulia-internal.1.dylib (unknown line)
jl_restore_incremental at /Users/mhorn/Applications/Julia-1.6.5.app/Contents/Resources/julia/lib/julia/libjulia-internal.1.dylib (unknown line)
_include_from_serialized at ./loading.jl:680
_require_from_serialized at ./loading.jl:749
_require at ./loading.jl:1053
require at ./loading.jl:936
require at ./loading.jl:923
jl_apply_generic at /Users/mhorn/Applications/Julia-1.6.5.app/Contents/Resources/julia/lib/julia/libjulia-internal.1.dylib (unknown line)
eval_import_path at /Users/mhorn/Applications/Julia-1.6.5.app/Contents/Resources/julia/lib/julia/libjulia-internal.1.dylib (unknown line)
jl_toplevel_eval_flex at /Users/mhorn/Applications/Julia-1.6.5.app/Contents/Resources/julia/lib/julia/libjulia-internal.1.dylib (unknown line)
jl_toplevel_eval_flex at /Users/mhorn/Applications/Julia-1.6.5.app/Contents/Resources/julia/lib/julia/libjulia-internal.1.dylib (unknown line)
jl_toplevel_eval_in at /Users/mhorn/Applications/Julia-1.6.5.app/Contents/Resources/julia/lib/julia/libjulia-internal.1.dylib (unknown line)
eval at ./boot.jl:360 [inlined]
eval_user_input at /Users/julia/buildbot/worker/package_macos64/build/usr/share/julia/stdlib/v1.6/REPL/src/REPL.jl:139
repl_backend_loop at /Users/julia/buildbot/worker/package_macos64/build/usr/share/julia/stdlib/v1.6/REPL/src/REPL.jl:200
start_repl_backend at /Users/julia/buildbot/worker/package_macos64/build/usr/share/julia/stdlib/v1.6/REPL/src/REPL.jl:185
#run_repl#42 at /Users/julia/buildbot/worker/package_macos64/build/usr/share/julia/stdlib/v1.6/REPL/src/REPL.jl:317
run_repl at /Users/julia/buildbot/worker/package_macos64/build/usr/share/julia/stdlib/v1.6/REPL/src/REPL.jl:305
jl_apply_generic at /Users/mhorn/Applications/Julia-1.6.5.app/Contents/Resources/julia/lib/julia/libjulia-internal.1.dylib (unknown line)
#875 at ./client.jl:387
jfptr_YY.875_23517.clone_1 at /Users/mhorn/Applications/Julia-1.6.5.app/Contents/Resources/julia/lib/julia/sys.dylib (unknown line)
jl_apply_generic at /Users/mhorn/Applications/Julia-1.6.5.app/Contents/Resources/julia/lib/julia/libjulia-internal.1.dylib (unknown line)
jl_f__call_latest at /Users/mhorn/Applications/Julia-1.6.5.app/Contents/Resources/julia/lib/julia/libjulia-internal.1.dylib (unknown line)
#invokelatest#2 at ./essentials.jl:708 [inlined]
invokelatest at ./essentials.jl:706 [inlined]
run_main_repl at ./client.jl:372
exec_options at ./client.jl:302
_start at ./client.jl:485
jfptr__start_30686.clone_1 at /Users/mhorn/Applications/Julia-1.6.5.app/Contents/Resources/julia/lib/julia/sys.dylib (unknown line)
jl_apply_generic at /Users/mhorn/Applications/Julia-1.6.5.app/Contents/Resources/julia/lib/julia/libjulia-internal.1.dylib (unknown line)
true_main at /Users/mhorn/Applications/Julia-1.6.5.app/Contents/Resources/julia/lib/julia/libjulia-internal.1.dylib (unknown line)
repl_entrypoint at /Users/mhorn/Applications/Julia-1.6.5.app/Contents/Resources/julia/lib/julia/libjulia-internal.1.dylib (unknown line)
Allocations: 29902483 (Pool: 29869107; Big: 33376); GC: 22

signal (11): Segmentation fault: 11
in expression starting at REPL[1]:1
jl_gc_alloc at /Users/mhorn/Applications/Julia-1.6.5.app/Contents/Resources/julia/lib/julia/libjulia-internal.1.dylib (unknown line)
_new_array_ at /Users/mhorn/Applications/Julia-1.6.5.app/Contents/Resources/julia/lib/julia/libjulia-internal.1.dylib (unknown line)
_new_array at /Users/mhorn/Applications/Julia-1.6.5.app/Contents/Resources/julia/lib/julia/libjulia-internal.1.dylib (unknown line)
jl_alloc_array_1d at /Users/mhorn/Applications/Julia-1.6.5.app/Contents/Resources/julia/lib/julia/libjulia-internal.1.dylib (unknown line)
Array at ./boot.jl:448 [inlined]
Array at ./boot.jl:457 [inlined]
similar at ./abstractarray.jl:785 [inlined]
similar at ./abstractarray.jl:784 [inlined]
_array_for at ./array.jl:672 [inlined]
_array_for at ./array.jl:670 [inlined]
IncrementalCompact at ./compiler/ssair/ir.jl:579
compact! at ./compiler/ssair/ir.jl:1395
compact! at ./compiler/ssair/ir.jl:1395 [inlined]
run_passes at ./compiler/ssair/driver.jl:139
optimize at ./compiler/optimize.jl:272 [inlined]
_typeinf at ./compiler/typeinfer.jl:244
typeinf at ./compiler/typeinfer.jl:209
abstract_call_method_with_const_args at ./compiler/abstractinterpretation.jl:330
abstract_call_gf_by_type at ./compiler/abstractinterpretation.jl:168
abstract_call_known at ./compiler/abstractinterpretation.jl:1048
abstract_call at ./compiler/abstractinterpretation.jl:1071
abstract_call at ./compiler/abstractinterpretation.jl:1055
abstract_eval_statement at ./compiler/abstractinterpretation.jl:1182
typeinf_local at ./compiler/abstractinterpretation.jl:1455
typeinf_nocycle at ./compiler/abstractinterpretation.jl:1534
_typeinf at ./compiler/typeinfer.jl:214
typeinf at ./compiler/typeinfer.jl:209
abstract_call_method_with_const_args at ./compiler/abstractinterpretation.jl:330
abstract_call_gf_by_type at ./compiler/abstractinterpretation.jl:168
abstract_call_known at ./compiler/abstractinterpretation.jl:1048
abstract_call at ./compiler/abstractinterpretation.jl:1071
abstract_apply at ./compiler/abstractinterpretation.jl:753
abstract_call_known at ./compiler/abstractinterpretation.jl:967
abstract_call at ./compiler/abstractinterpretation.jl:1071
abstract_call at ./compiler/abstractinterpretation.jl:1055
abstract_eval_statement at ./compiler/abstractinterpretation.jl:1182
typeinf_local at ./compiler/abstractinterpretation.jl:1470
typeinf_nocycle at ./compiler/abstractinterpretation.jl:1534
_typeinf at ./compiler/typeinfer.jl:214
typeinf at ./compiler/typeinfer.jl:209
abstract_call_method_with_const_args at ./compiler/abstractinterpretation.jl:330
abstract_call_gf_by_type at ./compiler/abstractinterpretation.jl:168
abstract_call_known at ./compiler/abstractinterpretation.jl:1048
abstract_call at ./compiler/abstractinterpretation.jl:1071
abstract_call at ./compiler/abstractinterpretation.jl:1055
abstract_eval_statement at ./compiler/abstractinterpretation.jl:1182
typeinf_local at ./compiler/abstractinterpretation.jl:1470
typeinf_nocycle at ./compiler/abstractinterpretation.jl:1534
_typeinf at ./compiler/typeinfer.jl:214
typeinf at ./compiler/typeinfer.jl:209
abstract_call_method_with_const_args at ./compiler/abstractinterpretation.jl:330
abstract_call_gf_by_type at ./compiler/abstractinterpretation.jl:168
abstract_call_known at ./compiler/abstractinterpretation.jl:1048
abstract_call at ./compiler/abstractinterpretation.jl:1071
abstract_call at ./compiler/abstractinterpretation.jl:1055
abstract_eval_statement at ./compiler/abstractinterpretation.jl:1182
typeinf_local at ./compiler/abstractinterpretation.jl:1470
typeinf_nocycle at ./compiler/abstractinterpretation.jl:1534
_typeinf at ./compiler/typeinfer.jl:214
typeinf at ./compiler/typeinfer.jl:209
typeinf_ext at ./compiler/typeinfer.jl:892
typeinf_ext_toplevel at ./compiler/typeinfer.jl:925
typeinf_ext_toplevel at ./compiler/typeinfer.jl:921
jfptr_typeinf_ext_toplevel_13611.clone_1 at /Users/mhorn/Applications/Julia-1.6.5.app/Contents/Resources/julia/lib/julia/sys.dylib (unknown line)
jl_apply_generic at /Users/mhorn/Applications/Julia-1.6.5.app/Contents/Resources/julia/lib/julia/libjulia-internal.1.dylib (unknown line)
jl_type_infer at /Users/mhorn/Applications/Julia-1.6.5.app/Contents/Resources/julia/lib/julia/libjulia-internal.1.dylib (unknown line)
jl_generate_fptr at /Users/mhorn/Applications/Julia-1.6.5.app/Contents/Resources/julia/lib/julia/libjulia-internal.1.dylib (unknown line)
jl_compile_method_internal at /Users/mhorn/Applications/Julia-1.6.5.app/Contents/Resources/julia/lib/julia/libjulia-internal.1.dylib (unknown line)
jl_apply_generic at /Users/mhorn/Applications/Julia-1.6.5.app/Contents/Resources/julia/lib/julia/libjulia-internal.1.dylib (unknown line)
_atexit at ./initdefs.jl:343
jfptr__atexit_43598.clone_1 at /Users/mhorn/Applications/Julia-1.6.5.app/Contents/Resources/julia/lib/julia/sys.dylib (unknown line)
jl_apply_generic at /Users/mhorn/Applications/Julia-1.6.5.app/Contents/Resources/julia/lib/julia/libjulia-internal.1.dylib (unknown line)
jl_atexit_hook at /Users/mhorn/Applications/Julia-1.6.5.app/Contents/Resources/julia/lib/julia/libjulia-internal.1.dylib (unknown line)
jl_exit at /Users/mhorn/Applications/Julia-1.6.5.app/Contents/Resources/julia/lib/julia/libjulia-internal.1.dylib (unknown line)
jl_exit_thread0_cb at /Users/mhorn/Applications/Julia-1.6.5.app/Contents/Resources/julia/lib/julia/libjulia-internal.1.dylib (unknown line)
Allocations: 29902483 (Pool: 29869107; Big: 33376); GC: 22
Segmentation fault: 11
```

I'll play a bit more to see if I can figure out where it breaks and report it, but in the meantime, that looks as if we need to do something different *sigh*

fingolfin commented:
It crashes with 1.6 and 1.7, but with 1.7 one gets an at least somewhat helpful error (which I added for this purpose :joy:): `Cannot serialize instances of foreign datatypes`

So the problem is the `GapObj` type from GAP.jl. Indeed, once I reverted the parts of my patches applying to the `src/Groups/` directory, it passes.

I'll investigate some more, as I don't quite understand why it would try to serialize an *instance* of that type, as opposed to just the type itself (which ought to be OK).

thofma commented:
@fingolfin Should we put this just in for the next release?

fingolfin commented:
I was still testing (then I got the new computer and all kinds of interruptions...). I'll try to finish those up tonight and we can then decide?

Although, even if it turns out to be a performance problem, we can downgrade transparently by removing the automatic return type detection from the macro (and replacing it by a simpler mechanism... I'll write more once I am on a computer not a phone)

thofma commented:
I don't mind putting it in. We don't have to use it if we encounter a problem of whatever kind.

fingolfin commented:
Then just go ahead :-)
